shark-ai/export_and_serve.py

```diff
@@ -0,0 +1,412 @@
+                "-m",
+                "sharktank.examples.sharding.shard_llm_dataset",
+                f"--irpa-file={self.weight_loc}",
+                f"--output-irpa={shard_file}",
+                "--tensor-parallelism-size=8",
+            ],
+            capture_output=True,
+        )
+
+        if shard_subp.returncode != 0:
+            logger.error("Failed to shard model to TP8")
+            print(shard_subp.stderr.decode(), file=sys.stderr)
+            exit(shard_subp.returncode)
+
+        logger.info(
+            f"Successfully sharded weight file; Unranked weight file located at {shard_file}"
+        )
+
+        # TODO(vinayakdsci): Refactor this into a separate function.
+        logger.info("Starting TP8 model export")
+        export_subp = subprocess.run(
+            [
+                "python3",
+                "-m",
+                "sharktank.examples.export_paged_llm_v1",
+                f"--irpa-file={shard_file}",
+                f"--output-mlir={str(self.mlir_path)}",
+                f"--output-config={self.model_config_path}",
+                "--bs=4",
+            ],
+            capture_output=True,
+        )
+
+        if export_subp.returncode != 0:
+            logger.error("Failed to export model in TP8 mode")
+            print(export_subp.stderr.decode(), file=sys.stderr)
+            exit(export_subp.returncode)
+
+        logger.info(
+            f"Successfully exported and saved model MLIR and config to {self.export_dir}"
+        )
+
+    def compile_model(self) -> None:
+        if not shutil.which("iree-compile"):
+            logger.error("iree-compile not found in PATH")
+            raise FileNotFoundError(
+                "iree-compile binary is required to be in PATH for compilation."
+            )
+
+        if self.shard or self.tp == 8:
+            self._compile_tp8()
+        else:
+            self._compile_tp1()
+
+    def _compile_tp1(self):
+        logger.info("Compiling unsharded IR")
+        tp1_vmfb_path = os.path.join(self.export_dir, "model.vmfb")
+        cmd = [
+            "iree-compile",
+            f"{self.mlir_path}",
+            f"-o={tp1_vmfb_path}",
+        ]
+
+        # TODO(vinayakdsci): Add a flag to support backends other than rocm.
+        cmd += [f"--iree-hal-target-backends={self.hal_target_backend}"]
+
+        # TODO(vinayakdsci): Add a flag to support targets other than gfx942.
+        cmd += [f"--iree-hip-target={self.hip_target}"]
+
+        compile_subp = subprocess.run(
+            cmd,
+            capture_output=True,
+        )
+
+        if compile_subp.returncode != 0:
+            logger.error("Failed to compile unsharded IR")
+            print(compile_subp.stderr.decode(), file=sys.stderr)
+            exit(compile_subp.returncode)
+
+        logger.info(
+            f"Successfully compiled and saved TP1 VMFB module to {tp1_vmfb_path}"
+        )
+
+    def _compile_tp8(self):
+        logger.info("Compiling sharded IR")
```

Comment:
May also want to include a time estimate here, or ideally a progress bar (https://github.com/iree-org/iree/issues/14369)